File: leetcode/c++/medium/_2966_divide_array_into_arrays_with_max_difference.cpp
// Leetcode Problem 2966

#include <gtest/gtest.h>

using namespace std;

class Solution {
public:
    vector<vector<int>> divideArray(vector<int>& nums, int k) {
        size_t n = nums.size();

        sort(nums.begin(), nums.end());

        vector<vector<int>> solution = {};
        for (int i = 0; i < n/3; i++) {
            if (nums[i*3+2] - nums[i*3] > k) {
                return {};
            }
            solution.push_back({nums[i*3], nums[i*3+1], nums[i*3+2]});
        }
        return solution;
    }
};


class TestDivideArray : public ::testing::Test {
protected:
    Solution solution;

    bool checkArray(vector<vector<int>> arr, int k) {
        for (vector<int> v : arr) {
            sort(v.begin(), v.end());
            if (v[2] - v[0] > k) {
                return false;
            }
        }
        return true;
    }
};

TEST_F(TestDivideArray, SimpleTestCase1) {
    vector<int> nums = {1,3,4,8,7,9,3,5,1};

    EXPECT_TRUE(this->checkArray(solution.divideArray(nums, 2), 2));
}

TEST_F(TestDivideArray, SimpleTestCase2) {
    vector<int> nums = {2,4,2,2,5,2};

    EXPECT_TRUE(this->checkArray(solution.divideArray(nums, 2), 2));
}

TEST_F(TestDivideArray, SimpleTestCase3) {
    vector<int> nums = {4,2,9,8,2,12,7,12,10,5,8,5,5,7,9,2,5,11};

    EXPECT_TRUE(this->checkArray(solution.divideArray(nums, 14), 14));
}

File: leetcode/rust/problems/src/medium/_2966_divide_array_into_arrays_with_max_difference.rs
// Leetcode Problem 2966

struct Solution;

#[allow(dead_code)]
impl Solution {
    pub fn divide_array(mut nums: Vec<i32>, k: i32) -> Vec<Vec<i32>> {
        nums.sort();

        let n = nums.len();
        let mut solution: Vec<Vec<i32>> = Vec::new();

        for i in 0..n/3 {
            if nums[i * 3 + 2] - nums[i * 3] > k {
                return vec![];
            }
            solution.push(vec![nums[i*3], nums[i*3+1], nums[i*3+2]]);
        }

        solution
    }
}


#[test]
fn test_divide_array() {
    let nums1 = vec![1, 3, 4, 8, 7, 9, 3, 5, 1];
    // let expected1 = vec![vec![1, 1, 3], vec![3, 4, 5], vec![7, 8, 9]];

    let nums2 = vec![2, 4, 2, 2, 5, 2];
    // let expected2: Vec<Vec<i32>> = vec![];
    
    let num3 = vec![4, 2, 9, 8, 2, 12, 7, 12, 10, 5, 8, 5, 5, 7, 9, 2, 5, 11];
    // let expected3 = vec![vec![2, 2, 12], vec![4, 8, 5], vec![5, 9, 7], vec![7, 8, 5], vec![5, 9, 10], vec![11, 12, 2]];

    assert!(check_array(&mut Solution::divide_array(nums1, 2), 2));
    assert!(check_array(&mut Solution::divide_array(nums2, 2), 2));
    assert!(check_array(&mut Solution::divide_array(num3, 14), 14));
}

#[allow(dead_code)]
fn check_array(arr: &mut Vec<Vec<i32>>, k: i32) -> bool {
    for v in arr {
        v.sort();
        if v[2] - v[0] > k {
            return false;
        }
    }
    true
}